### ISSUE TITLE
Open network XML returns unformatted

### ISSUE THREAD

**Issue by akinuri** [jorgebv/windows-wifi-api#4]
Using the app I'm building, I add three networks with different security settings.

![wnm](https://cloud.githubusercontent.com/assets/13146977/8591656/7a84f742-2632-11e5-9880-a3052c755fd3.PNG)

Later I export the network profiles in XML format.

![profiles-folder](https://cloud.githubusercontent.com/assets/13146977/8591683/a3c193e0-2632-11e5-84c8-035945c92f4f.PNG)

XML format of wep and wpa networks looks formatted (indented)

![wepnetwork](https://cloud.githubusercontent.com/assets/13146977/8591721/ed343370-2632-11e5-92f9-de939650ccec.PNG)

but not the open network.

![opennetwork](https://cloud.githubusercontent.com/assets/13146977/8591734/f960681c-2632-11e5-8ee9-9984e9925c31.PNG)

So I started investigating. Found the actual folder where the XML files are stored which is

`C:\ProgramData\Microsoft\Wlansvc\Profiles\Interfaces\{2B582EE6-A8F8-4122-AA09-300969E3B9D6}`

![wlan-folder](https://cloud.githubusercontent.com/assets/13146977/8591846/aad3a366-2633-11e5-933d-7ec67363d7d6.PNG)

Found and opened the same XML file that was returned unformatted (not indented)

![ms-open-network](https://cloud.githubusercontent.com/assets/13146977/8591866/cfd35b34-2633-11e5-865e-667fdb8a4b02.PNG)

It was indented! Why didn't managed-wlan-api return the the xml indented? I got more curious. I added an open network using Windows's tool.

![ms-new-open-network](https://cloud.githubusercontent.com/assets/13146977/8592053/121ded82-2635-11e5-9b14-aa0746769067.PNG)

Then checked the XML in the Interfaces folder. It was indented. Okay, that's expected.

![ms-open-xml](https://cloud.githubusercontent.com/assets/13146977/8592104/64d27b6a-2635-11e5-9d60-acb1d6cabec5.PNG)

Then exported the same XML file using my app.

![ms-open](https://cloud.githubusercontent.com/assets/13146977/8592167/d299ef16-2635-11e5-9b46-6ae332440adb.PNG)

It was indented. How come?

The code I'm using to turn user input to an XML file is:

`profileXML = string.Format("<?xml version=\"1.0\"?> <WLANProfile xmlns=\"http://www.microsoft.com/networking/WLAN/profile/v1\"> <name>{0}</name> <SSIDConfig> <SSID> <hex>{1}</hex> <name>{0}</name> </SSID> <nonBroadcast>{2}</nonBroadcast> </SSIDConfig> <connectionType>ESS</connectionType> <connectionMode>{3}</connectionMode> <MSM> <security> <authEncryption> <authentication>{4}</authentication> <encryption>{5}</encryption> <useOneX>false</useOneX> </authEncryption> <sharedKey> <keyType>{6}</keyType> <protected>false</protected> <keyMaterial>{7}</keyMaterial> </sharedKey> </security> </MSM> </WLANProfile>", networkName, networkNameHex, nonBroadcast, connectionMode, securityType, encryptionType, keyType, securityKey);`

Open networks' XML structure is different than weps' and wpas', so it has a different code:

`profileXML = string.Format("<?xml version=\"1.0\"?> <WLANProfile xmlns=\"http://www.microsoft.com/networking/WLAN/profile/v1\"> <name>{0}</name> <SSIDConfig> <SSID> <hex>{1}</hex> <name>{0}</name> </SSID> <nonBroadcast>{2}</nonBroadcast> </SSIDConfig> <connectionType>ESS</connectionType> <connectionMode>{3}</connectionMode> <MSM> <security> <authEncryption> <authentication>{4}</authentication> <encryption>none</encryption> <useOneX>false</useOneX> </authEncryption> </security> </MSM> </WLANProfile>", networkName, networkNameHex, nonBroadcast, connectionMode, securityType);`

Indented formats of the profileXML vars:

``` csharp
string.Format("<?xml version=\"1.0\"?>
<WLANProfile xmlns=\"http://www.microsoft.com/networking/WLAN/profile/v1\">
    <name>{0}</name>
    <SSIDConfig>
        <SSID>
            <hex>{1}</hex>
            <name>{0}</name>
        </SSID>
        <nonBroadcast>{2}</nonBroadcast>
    </SSIDConfig>
    <connectionType>ESS</connectionType>
    <connectionMode>{3}</connectionMode>
    <MSM>
        <security>
            <authEncryption>
                <authentication>{4}</authentication>
                <encryption>{5}</encryption>
                <useOneX>false</useOneX>
            </authEncryption>
            <sharedKey>
                <keyType>{6}</keyType>
                <protected>false</protected>
                <keyMaterial>{7}</keyMaterial>
            </sharedKey>
        </security>
    </MSM>
</WLANProfile>", networkName, networkNameHex, nonBroadcast, connectionMode, securityType, encryptionType, keyType, securityKey);
```

for open networks:

``` csharp
string.Format("<?xml version=\"1.0\"?>
<WLANProfile xmlns=\"http://www.microsoft.com/networking/WLAN/profile/v1\">
    <name>{0}</name>
    <SSIDConfig>
        <SSID>
            <hex>{1}</hex>
            <name>{0}</name>
        </SSID>
        <nonBroadcast>{2}</nonBroadcast>
    </SSIDConfig>
    <connectionType>ESS</connectionType>
    <connectionMode>{3}</connectionMode>
    <MSM>
        <security>
            <authEncryption>
                <authentication>{4}</authentication>
                <encryption>none</encryption>
                <useOneX>false</useOneX>
            </authEncryption>
        </security>
    </MSM>
</WLANProfile>", networkName, networkNameHex, nonBroadcast, connectionMode, securityType);
```

To sum it up
- Windows creates (in Interfaces folder) and exports (using netsh cmd) open network xml files indented
- managed-wifi-api (or my app [xml strings]) creates open network xml files (in Interfaces folder)  indented, but exports them unindented (as a single line)
- managed-wifi-api exports the open network xml files created with windows' tool indented

So my final conclusion is that either the way I'm building the XML string for open networks is problematic, or managed-wifi-api messes the XML in the process.

<hr>

Sorry this got really long. It's not really a problem, so it's okay if there isn't a way to solve this. I'm a curious person, I just wanted to know.

**Comment by jorgebv** [comment]
This does seem a bit unusual, so I looked into netsh, wlansvc, and a few other things to figure out what was going on and why it is inconsistent. Here is what I found.
1. Even though your ms-open network appears formatted in the interfaces folder, it still is unindented in memory inside of wlansvc. It is indented when it is written to disk, but that doesn't seem to be persisted.
2. Your wepNetwork is indented because wlansvc needs to regenerate the XML in order to encrypt the key. You pass in an unencrypted key, but wlansvc will encrypt it and generate a new XML that it keeps in memory and on disk. This XML is indented. Take a look at the XML inside the interfaces folder -- you'll see an encrypted version of the XML you passed in.
3. When you export with netsh, it is formatting the profiles for you. It is calling WlanGetProfile and then formatting them using a DOMDocument60.Load function (note that this is a COM class that is available through a C# interop, but I don't recommend using it. If you want to work with XML I recommend the System.Xml namespace)

I don't know how clear the above summary is. But the end result is:
- The managed-wifi-api is returning exactly what the native API returns
- Secure networks are indented because wlansvc needs to regenerate the XML in order to encrypt the key
- Open networks are returned exactly as you passed them in
- If you want a formatting guarantee, you should format the networks as you pass/set them in, or after you get them back. Since you don't have 100% control over created networks, I would recommend formatting after getting them.

Let me know if you have any questions or if anything is unclear!

**Comment by jorgebv** [comment]
One more note -- open networks are returned exactly as you passed them in because they are still in memory. If you shut down wlansvc and restart it, they will be unloaded from memory, and then they will be reloaded from disk. After this, they will be indented (because the copy in the interface folder is indented)!

You can restart wlansvc by opening a command prompt and running:
net stop wlansvc
net start wlansvc

or restarting you computer.

**Comment by akinuri** [comment]
I did what you suggested. Formatted the xml string after I got it. The function is taken from Neil C. Obremski's [answer](http://stackoverflow.com/a/203581/2202732).

``` csharp
public static string BeautifyXML(XmlDocument doc)
{
    StringBuilder sb = new StringBuilder();
    XmlWriterSettings settings = new XmlWriterSettings
    {
        Indent = true,
        IndentChars = "\t",
        NewLineChars = "\r\n",
        NewLineHandling = NewLineHandling.Replace
    };
    using (XmlWriter writer = XmlWriter.Create(sb, settings))
    {
        doc.Save(writer);
    }
    return sb.ToString();
}
```

And used it like this

``` csharp
string selectedProfile = ProfilesListView.SelectedItems[0].SubItems[1].Text;
string profileXML = "";

WlanClient client = new WlanClient();

foreach (WlanClient.WlanInterface wlanIface in client.Interfaces)
{
    profileXML = wlanIface.GetProfileXml(selectedProfile, true);
}

XmlDocument doc = new XmlDocument();
doc.LoadXml(profileXML);

profileXML = Tools.BeautifyXML(doc);
```

It's working great. Thanks again.
